### PR TITLE
Oauth refactor

### PR DESCRIPTION
**Pull request by arapov**
1. Once we know more about CRDs, and OAuth settings for OCPv4 specifically, we want to generalize OCP4 Oauth code.

2. Use example configuration yaml to run app.
2.1. Use hostname as a *Source* in yaml for the first time.
2.2. Replace hostname in *Source* by local directory, so you don't wait for SSH everytime.

This is robust prototype, feel free to improve.

### CODE REVIEW

File: ocp4crd/oauth/oauth.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+	configv1 "github.com/openshift/api/legacyconfig/v1"
+	oauthv1 "github.com/openshift/api/oauth/v1"
+	"gopkg.in/yaml.v2"
+
+	"k8s.io/apimachinery/pkg/runtime/serializer/json"
+	"k8s.io/client-go/kubernetes/scheme"
+)
+
+func init() {
+	configv1.InstallLegacy(scheme.Scheme)
```
Comment (arapov) [inline ocp4crd/oauth/oauth.go]: This one is not needed here. Though let it be here for now and remove in future commits.

File: ocp4crd/oauth/oauth.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+// Generate converts OCPv3 OAuth to OCPv4 OAuth Custom Resources
+func Generate(masterconfig configv1.MasterConfig) error {
+	var auth = masterconfig.OAuthConfig
+	var err error
+
+	var crd v4OAuthCRD
+	crd.APIVersion = "config.openshift.io/v1"
+	crd.Kind = "OAuth"
+	crd.MetaData.Name = "cluster"
+
```
Comment (arapov) [inline ocp4crd/oauth/oauth.go]: This stuff above, I guess, is possible to decipher from the original master-config and get rid of hard coding.

File: examples/cpma-config.example.yaml [hunk 3]
```diff
@@ -1,6 +1,6 @@
 ---
-Source:
-  UserName: "root"
-  HostName: "master-0.gildub1.lab.pnq2.cee.redhat.com"
-  SSHKey: "/home/gildub/.ssh/quicklab"
-outputPath: "./data"
\ No newline at end of file
+Source: "master-0.gildub1.lab.pnq2.cee.redhat.com" # can also point to directory
```
Comment (arapov) [inline examples/cpma-config.example.yaml]: After the first run you can make Source equal to OutputDir and CPMA will be working with the local configuration files copy.

File: env/env.go [hunk 4]
```diff
@@ -1,334 +1,17 @@
...
-	"encoding/json"
-	"fmt"
-	"html/template"
-	"io/ioutil"
-	"os"
-	"path/filepath"
-	"reflect"
-	"strings"
-
+	"github.com/davecgh/go-spew/spew"
```
Comment (arapov) [inline env/env.go]: spew is temporary. this is TODO: to figure out good logger which can pretty print structures.

File: cmd/root.go [hunk 5]
```diff
@@ -48,13 +48,9 @@ func Execute() {
 func init() {
 	cobra.OnInitialize(config.InitConfig)
 	rootCmd.PersistentFlags().StringVar(&config.ConfigFile, "config", "", "config file (default is $HOME/.cpma.yaml)")
-	rootCmd.PersistentFlags().BoolVar(&debugLogLevel, "debug", false, "set log level to debug")
+	rootCmd.PersistentFlags().BoolVar(&debugLogLevel, "debug", false, "show debug ouput")
```
Comment (arapov) [inline cmd/root.go]: help, more useful for end user

File: cmd/root.go [hunk 6]
```diff
@@ -48,13 +48,9 @@ func Execute() {
...
 
 	rootCmd.Flags().StringP("output-dir", "o", "", "set the directory to store extracted configuration.")
-	config.Config().BindPFlag("outputPath", rootCmd.Flags().Lookup("output-dir"))
-	config.Config().SetDefault("outputPath", path.Dir(""))
-
-	rootCmd.Flags().BoolP("local-only", "l", false, "do not fetch files, use only local files.")
```
Comment (arapov) [inline cmd/root.go]: we don't want to abuse CPMA with parameters. See the comment to example-config below.

File: ocp4crd/oauth/oauth.go [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+			idP.Challenge = p.UseAsChallenger
+			idP.Login = p.UseAsLogin
+			idP.MappingMethod = p.MappingMethod
+			idP.GitHub.HostName = github.Hostname
+			idP.GitHub.CA.Name = github.CA
+			idP.GitHub.ClientID = github.ClientID
+			idP.GitHub.Organizations = github.Organizations
+			idP.GitHub.Teams = github.Teams
+			// TODO: Learn how to handle secrets
+			idP.GitHub.ClientSecret.Name = github.ClientSecret.Value
```
Comment (arapov) [inline ocp4crd/oauth/oauth.go]: Secrets can be in different places: ENV var, configmap, secrets, plaintext. We have to learn how to handle those.

File: ocp4crd/oauth/oauth.go [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+		}
+
+	}
+
+	yamlBytes, err := yaml.Marshal(&crd)
+	if err != nil {
+		log.Fatal(err)
+	}
+	log.Println(string(yamlBytes))
+	return err
```
Comment (arapov) [inline ocp4crd/oauth/oauth.go]: This printout CRD yaml. We rather need to implement proper file writers. Also, I'm not sure if we want to split various providers into multiple CR files.

**Comment by alexander-demichev** [comment]
Something seems to be broken here:
```
ademicev-OSX:cpma ademicev$ ./bin/cpma --config /Users/ademicev/Documents/go/src/github.com/fusor/cpma/examples/cpma-config.yaml
36/40 10:36:52 FATAL unable to parse configuration: 1 error(s) decoding:

* 'Source' expected type 'string', got unconvertible type 'map[string]interface {}'
```

File: ocp4crd/oauth/oauth.go (hunk 2 above)
Comment (gildub) [reply ocp4crd/oauth/oauth.go]: Yes that's the plan.

**Comment by alexander-demichev** [comment]
Do we want to move `ocp3` and `ocp4` packages under `internal` dir?

**Comment by alexander-demichev** [comment]
> Do we want to move `ocp3` and `ocp4` packages under `internal` dir?

Works fine, didn't notice you changed config

File: cmd/root.go (hunk 6 above)
Comment (gildub) [reply cmd/root.go]: +1

File: ocp4crd/oauth/oauth.go (hunk 8 above)
Comment (gildub) [reply ocp4crd/oauth/oauth.go]: The way OCP4 works is by components, so we're likely to break it down the same way.

**Comment by arapov** [comment]
This implements #25

**Pull request merged**